### PR TITLE
Add websocket stream

### PR DESCRIPTION
**Pull request by Bishoy-at-pieces**

### CODE REVIEW

File: assets/utils.py [hunk 1]
```diff
@@ -1,68 +1,10 @@
...
-					cls.asset_set.add(asset_id)  # Add asset_id to the set
-		cls.first_shot = False
-		cls.block = False # Remove the block to end the thread
-
-
-
-
-
-
+class AssetSnapshot(StreamedIdentifiersCache,
```
Comment (caleb-at-pieces) [inline assets/utils.py:5]: do we still need this class?

File: copilot/ask_view.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+		ui = sublime.ui_info()["theme"]["style"]
+
+		self.copilot_regions.append(sublime.Region(self.gpt_view.size(), self.gpt_view.size()))
+
+		# Add the regions with the icon and appropriate flags
+		self.gpt_view.add_regions(
+			"pieces", 
+			self.copilot_regions, 
+			scope="text", 
+			icon=f"Packages/Pieces/copilot/images/copilot-icon-{ui}.png",
```
Comment (caleb-at-pieces) [inline copilot/ask_view.py:99]: this works on windows?

File: copilot/ask_view.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+		self.last_edit_phantom = view.size()
+
+	def on_nav(self,href:str):
+		command,id = href.split("_")
+		code = self.phantom_details_dict[id]["code"]
+		region = self.phantom_details_dict[id]["region"]
+		if command == "save":
+			self.gpt_view.run_command("pieces_create_asset",{"data":code})
+			self.update_phantom_set(region,id,"Saving")
+			sublime.set_timeout_async(lambda:self.update_phantom_set(region,id,"Saved"),5000)
```
Comment (caleb-at-pieces) [inline copilot/ask_view.py:221]: how come we are using these timeouts here?

File: copilot/ask_view.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+		
+		if not reset:
+			phantoms = [phantom for phantom in self.phantom_set.phantoms if phantom.region != region]
+			phantoms = [phantom,*phantoms]
+		else: 
+			phantoms = [phantom]
+		self.phantom_set.update(phantoms)
+
+	@staticmethod
+	def render_copilot_image_phantom(view:sublime.View):
```
Comment (caleb-at-pieces) [inline copilot/ask_view.py:246]: can we just delete this method?

File: copilot/ask_view.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+		self.view_name = conversation.name
+		self.gpt_view.run_command("select_all")
+		self.gpt_view.run_command("right_delete") # Clear the cursor created by default ">>>"
+		message_api = ConversationMessageApi(PiecesSettings.api_client)
+
+		for key,val in conversation.messages.indices.items():
+			self.select_end
+			if val == -1: # message is deleted
+				continue
+			message = message_api.message_specific_message_snapshot(message=key)
```
Comment (caleb-at-pieces) [inline copilot/ask_view.py:284]: unsure, but we might need transferables: true here

File: copilot/ask_websocket.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	def on_message(self,ws, message):
+		self.on_message_callback(QGPTStreamOutput.from_json(message))
+
+	
+	def send_message(self,message:QGPTStreamInput):
+		try:
+			if not self.ws:
+				raise WebSocketConnectionClosedException()
+			self.ws.send(message.to_json())
+		except WebSocketConnectionClosedException:
```
Comment (caleb-at-pieces) [inline copilot/ask_websocket.py:26]: what does the error handling look like here when this fails as a user?

File: streamed_identifiers.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import threading
+from typing import Dict, Union, Callable
+from pieces_os_client import Conversation, StreamedIdentifiers, Asset
+import sublime
+
+
+class StreamedIdentifiersCache:
+    """
+    This class is made for caching Streamed Identifiers.
+    Please use this class only as a parent class.
```
Comment (caleb-at-pieces) [inline streamed_identifiers.py:34]: it looks like there is some packages out there for python that will ensure this is only used as an abstract class, consider implementing them

File: streamed_identifiers.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        cls.identifiers_snapshot: Dict[str, Union[Asset, Conversation,None]] = {}  # Map id:return from the api_call
+        cls.identifiers_queue = queue.Queue()  # Queue for ids to be processed
+        cls.identifiers_set = set()  # Set for ids in the queue
+        cls.api_call = api_call
+        cls.block = True  # to wait for the queue to receive the first id
+        cls.first_shot = True  # First time to open the websocket or not
+
+
+    @classmethod
+    def sort_first_shot(cls):
```
Comment (caleb-at-pieces) [inline streamed_identifiers.py:48]: do we need this method?

File: assets/utils.py (hunk 1 above)
Comment (Bishoy-at-pieces) [reply assets/utils.py:5]: Yes we will need it

File: copilot/ask_view.py (hunk 2 above)
Comment (Bishoy-at-pieces) [reply copilot/ask_view.py:99]: yes no worries these paths are handled by sublime

File: copilot/ask_websocket.py (hunk 6 above)
Comment (Bishoy-at-pieces) [reply copilot/ask_websocket.py:26]: nothing happens here it just reconnect to the websocket if it is not

File: streamed_identifiers.py (hunk 8 above)
Comment (Bishoy-at-pieces) [reply streamed_identifiers.py:48]: yupp we need it when we inherit sometimes I need to preform some sorts to sort the conversations for example but in the assets I don't need to sort them.

this class for inheritance only!

**Review by caleb-at-pieces — APPROVED** [review]

**Review by mark-at-pieces — APPROVED** [review]

**Pull request merged**